src/rgw/rgw_rest_s3.h

```diff
@@ -914,42 +914,129 @@ class s3select;
 class csv_object;
 }
 
+
+class aws_response_handler
+{//TODO this class should reside on s3select submodule 
+
+private:
+  std::string sql_result;
+  struct req_state *s;//TODO will be replace by callback
+  uint32_t header_size;
+  std::unique_ptr<boost::crc_32_type> crc32;
+  RGWOp *m_rgwop;
+  std::string m_buff_header;
+  uint64_t total_bytes_returned;
+  uint64_t processed_size;
+
+  enum header_name_En
+  {
+    EVENT_TYPE,
+    CONTENT_TYPE,
+    MESSAGE_TYPE,
+    ERROR_CODE,
+    ERROR_MESSAGE
+  };
+
+  enum header_value_En
+  {
+    RECORDS,
+    OCTET_STREAM,
+    EVENT,
+    CONT,
+    PROGRESS,
+    END,
+    XML,
+    STATS,
+    ENGINE_ERROR,
+    ERROR_TYPE
+  };
+
+  const char *PAYLOAD_LINE= "\n<Payload>\n<Records>\n<Payload>\n";
+  const char *END_PAYLOAD_LINE= "\n</Payload></Records></Payload>";
+  const char *header_name_str[5] =  {":event-type", ":content-type", ":message-type",":error-code",":error-message"};
+  const char *header_value_str[10] = {"Records", "application/octet-stream", "event", "Cont", "Progress", "End", "text/xml", "Stats", "s3select-engine-error","error"};
+
+  void push_header(const char * header_name,const char* header_value);
+
+  int create_message(u_int32_t header_len);
+
+public:
+  //12 positions for header-crc
+  aws_response_handler(struct req_state *ps,RGWOp *rgwop) : sql_result("012345678901"), s(ps),m_rgwop(rgwop),total_bytes_returned{0},processed_size{0}
+  {
+    // the parameters are according to CRC-32 algorithm and its aligned with AWS-cli checksum
+    crc32 = std::unique_ptr<boost::crc_32_type>(new boost::crc_optimal<32, 0x04C11DB7, 0xFFFFFFFF, 0xFFFFFFFF, true, true>);
```

Comment:
is it really necessary to heap-allocate here? 

from the boost docs at https://www.boost.org/doc/libs/develop/doc/html/crc/reference.html#header.boost.crc_hpp, i see that this `crc_optimal` type is the same as `crc_32_type`:
```c++
typedef crc_optimal< 32, 0x04C11DB7, 0xFFFFFFFF, 0xFFFFFFFF, true, true > crc_32_type;
```
so you should be able to just store `boost::crc_32_type crc32;` as the member variable. then the constructor doesn't need to do anything extra

Reply:
yes,
if it's not necessary (heap allocation), it will be removed.